Undeprecate the Traversal API and bring back old examples (#39)
neo4j/docs-java-reference: modules/ROOT/pages/traversal-framework/bidirectional_traversal_framework.adoc

td.traverse(startNode, endNode);
----

== Collision Policies
A `BranchCollisionPolicy` defines when a collision is detected and accepted in a bidirectional traversal. Given an evaluator and a path predicate,
a `BranchCollisionPolicy` will create a `BranchCollisionDetector`, which will detect collisions between two traversers and
add the resulting path to the result if it satisfies the conditions given by the collision evaluator and uniqueness constraints.

These are the build-in `BranchCollisionPolicies`:

* `STANDARD` -- Returns all paths with a collision.
* `SHORTEST_PATH` -- Returns all paths with the smallest traversal depth.

Here's an example of how to use a built-in `BranchCollisionPolicy`:
[source, java, role="nocopy"]
----
BidirectionalTraversalDescription td = transaction
        .bidirectionalTraversalDescription()
        .mirroredSides(transaction
                               .traversalDescription()
                               .uniqueness(Uniqueness.RELATIONSHIP_GLOBAL))
        .collisionPolicy(BranchCollisionPolicies.SHORTEST_PATH);
td.traverse(startNode, endNode);
----

== Collision Evaluator

> There is a convenience method `collisionEvaluator()` for adding a `PathEvaluator` which will validate paths in a valid collision.

Reviewer: Convenience methods is a technical term, so I would prefer keeping this.

:description: The Neo4j Bidirectional Traversal Framework Java API.

:org-neo4j-graphdb-bidirectional-traversal-description: {neo4j-javadocs-base-uri}/org/neo4j/graphdb/Transaction.html#bidirectionalTraversalDescription()

[[Bidirectional-Traversal-Framework]]
= The Bidirectional Traversal Framework

> A bidirectional traversal consists of two traversals starting from different nodes, returning the paths where both traversals collide.

Reviewer 2: Is there somewhere an image to show this? I think it could be useful to exemplify more visually what that means.

Reviewer 3: Even better, an example to go with the picture.

> The Bidirectional Traversal Framework allows the description of such a traversal with the link:{org-neo4j-graphdb-bidirectional-traversal-description}[BidirectionalTraversalDescription^].

Reviewer 2: Shouldn't these words be separated with spaces? 
```suggestion
The Bidirectional Traversal Framework allows the description of such a traversal with the link:{org-neo4j-graphdb-bidirectional-traversal-description}[Bidirectional Traversal Description^].
```

Author: This is the class name and it links to the class java docs

[NOTE]
====

> The Uniqueness is shared between both traversals, meaning there will be no results if the Uniqueness is set to `Uniqueness.NODE_GLOBAL`,

Reviewer 2: ```suggestion
The Uniqueness is shared between both traversals, which means that there will be no results if the Uniqueness is set to `Uniqueness.NODE_GLOBAL`,
```

as both traversals need to reach the same node to cause a collision.

> As a result of this, always set the uniqueness manually when using the `BidirectionalTraversalDescription` as the default is `Uniqueness.NODE_GLOBAL`.

Reviewer 2: ```suggestion
For this reason, always set the Uniqueness manually when using the `BidirectionalTraversalDescription`, since the default is `Uniqueness.NODE_GLOBAL`.
```

====

[source, java, role="nocopy"]
----
BidirectionalTraversalDescription td = transaction
        .bidirectionalTraversalDescription()
        .mirroredSides(transaction
                               .traversalDescription()
                               .uniqueness(Uniqueness.RELATIONSHIP_GLOBAL));
td.traverse(startNode, endNode);
----

== Collision Evaluator

> The `CollisionEvaluator` takes a `PathEvaluator`, which evaluates if a path is accepted as a result or not.

Reviewer 2: ```suggestion
The `CollisionEvaluator` takes a `PathEvaluator` that checks if a path is accepted as a result or not.
```

> It is possible to define multiple collision evaluators.

Reviewer 4: It's not clear to me what the Collision Evaluator is. It appears here and never more?

== Side Selector

> In a bidirectional traversal the traverser alternates which side (start or end) to move further on for each step.

Reviewer 2: Not sure if I changed the meaning here, but I was trying to make the sentence more simple. 

```suggestion
In a bidirectional traversal, the traverser selects which side (start or end) to move on each step.
```

> It is possible to change how this works by implementing the `SideSelectorPolicy` interface, which contains a function for

Reviewer 2: ```suggestion
It is possible to change how this works by implementing the `SideSelectorPolicy` interface, as it has a function for
```

determining which side to traverse the next step from.

* `Direction.OUTGOING` -- The traverser coming from the start node.
* `Direction.INCOMING` -- The traverser coming from the end node.

The built-in policies include:

* `SideSelectorPolicies.LEVEL` -- Stops traversal if the combined depth is larger than the given maximum depth.

> * `SideSelectorPolicies.LEVEL_STOP_DESCENT_ON_RESULT` -- Stops as soon as a result was found.

Reviewer 2: ```suggestion
* `SideSelectorPolicies.LEVEL_STOP_DESCENT_ON_RESULT` -- Stops as soon as a result is found.
```

* `SideSelectorPolicies.ALTERNATING` -- Alternates which branch continues the traversal.

> The `SideSelectorPolicy` optionally takes a maxDepth which represents the combined depth that both sides must adhere to.

Reviewer 2: ```suggestion
The `SideSelectorPolicy` optionally takes a `maxDepth` which represents the combined depth that both sides must adhere to.
```

=== Example

> Here's an example of using a built-in `SideSelectorPolicy` with a max combined depth of 5.

Reviewer 2: ```suggestion
Here's an example of how to use a built-in `SideSelectorPolicy` with a max combined depth of 5.
```

[source, java, role="nocopy"]
----
BidirectionalTraversalDescription td = transaction
        .bidirectionalTraversalDescription()
        .mirroredSides(transaction
                               .traversalDescription()
                               .uniqueness(Uniqueness.RELATIONSHIP_GLOBAL))
                               .sideSelector(SideSelectorPolicies.LEVEL, 5);
td.traverse(startNode, endNode);
----

== Collision Policies
The `BranchCollisionDetector` returns the paths which can be combined from the two given traversers.
A `BranchCollisionPolicy` then evaluates which of these paths should be included.

These are the build-in `BranchCollisionPolicies`:

* `STANDARD` -- Returns all paths with a collision.
* `SHORTEST_PATH` -- Returns all paths with the smallest traversal depth.

> Here's an example of using a built-in `BranchCollisionPolicy`:

Reviewer 2: ```suggestion
Here's an example of how to use a built-in `BranchCollisionPolicy`:
```
